pkg/maintenance/postgresql.go

```diff
@@ -0,0 +1,285 @@
+package maintenance
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/go-version"
+	"github.com/spf13/viper"
+	stackgresv1 "github.com/vshn/appcat/apis/stackgres/v1"
+	vshnv1 "github.com/vshn/appcat/apis/vshn/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"k8s.io/utils/strings/slices"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// PostgreSQL handles the maintenance of postgresql services
+type PostgreSQL struct {
+	Client            client.Client
+	log               logr.Logger
+	instanceNamespace string
+	ctx               context.Context
+	apiUserName       string
+	apiPassword       string
+	claimNamespace    string
+	claimName         string
+	SgURL             string
+}
+
+type loginRequest struct {
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
+}
+
+type authToken struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"`
+}
+
+type pgVersions struct {
+	Postgresql []string `json:"postgresql"`
+}
+
+// DoMaintenance will run postgresql's maintenance script.
+func (p *PostgreSQL) DoMaintenance(ctx context.Context) error {
+
+	if err := p.configure(); err != nil {
+		return err
+	}
+
+	p.ctx = ctx
+
+	p.log = logr.FromContextOrDiscard(p.ctx).WithValues("instanceNamespace", p.instanceNamespace)
+
+	p.log.Info("Starting maintenance on postgresql instance")
+
+	sgclusters, err := p.listClustersInNamespace()
+	if err != nil {
+		return err
+	}
+
+	if len(sgclusters.Items) == 0 {
+		p.log.Info("No sgcluster found in namespace, skipping maintenance")
+	}
+
+	clusterName := sgclusters.Items[0].GetName()
+	currentVersion := sgclusters.Items[0].Spec.Postgres.Version
+
+	versionList, err := p.fetchVersionList(p.SgURL)
+	if err != nil {
+		p.log.Error(err, "StackGres API error")
+		p.log.Info("Can't get latest minor version, proceeding with security maintenance")
+	}
+
+	// if there are any errors here, we fall back to a security upgrade
+	latestMinor, err := p.getLatestMinorversion(currentVersion, versionList)
+	if err != nil {
+		p.log.Error(err, "Could not get latest minor version from list, continuing with security upgrade")
+		currentVersion = latestMinor
+	}
+
+	p.log.Info("Found versions", "current", currentVersion, "latest", latestMinor)
+	if currentVersion != latestMinor {
+		p.log.Info("Doing a minor upgrade")
+		return p.createMinorUpgrade(clusterName, latestMinor)
+	}
+
+	p.log.Info("Checking for EOL")
+	if versionList != nil && p.isEOL(currentVersion, *versionList) {
```

Comment:
Is this related with the operator update? I mean the version list will remain always the same unless we upgrade the operator?

Reply:
Yes, if we update the operator then it will set the EOL accordingly. We currently don't send mails for that.